### PR TITLE
Speed up ByteBuffer setBytes slow path

### PR DESCRIPTION
**Pull request by richardalow**
Add a benchmark for CircularBuffer<UInt8> setBytes/writeBytes, see #1216.

Motivation:

When calling ByteBuffer setBytes/writeBytes with CircularBuffer<Uint8> we take the slow path. Before we make any changes we want a baseline for current performance.

Modifications:

Adds ByteBuferBenchmark.swift and tests circular_buffer_into_byte_buffer_1kb, circular_buffer_into_byte_buffer_1mb.

Result:

Extra performance tests

Don't merge yet until a baseline test has run.

### CODE REVIEW

**Comment by richardalow** [comment]
@swift-nio-bot test perf please

**Comment by Lukasa** [comment]
@swift-nio-bot add to whitelist

**Comment by Lukasa** [comment]
@swift-nio-bot test perf please

File: Sources/NIOPerformanceTester/ByteBufferBenchmark.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+
+final class ByteBufferBenchmark: Benchmark {
+    private let iterations : Int
+    private let bufferSize : Int
```
Comment (Lukasa) [inline Sources/NIOPerformanceTester/ByteBufferBenchmark.swift]: Extremely minor nit: NIO style is not to have a space before the colon.

File: Sources/NIOPerformanceTester/ByteBufferBenchmark.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    private var circularBuffer: CircularBuffer<UInt8>
+    private var buffer: ByteBuffer
+
+    init(iterations: Int, bufferSize: Int) {
+        self.iterations = iterations
+        self.bufferSize = bufferSize
+        self.circularBuffer = CircularBuffer<UInt8>(initialCapacity: self.bufferSize)
+        self.buffer = ByteBufferAllocator().buffer(capacity: self.bufferSize)
```
Comment (Lukasa) [inline Sources/NIOPerformanceTester/ByteBufferBenchmark.swift]: Nit: we should probably reserve `bufferSize * iterations` space, given that we call `writeBytes` in the loop. Alternatively, the loop could call `clear()` on each iteration.

Comment (richardalow) [reply Sources/NIOPerformanceTester/ByteBufferBenchmark.swift]: I added clear()

**Comment by richardalow** [comment]
@swift-nio-bot test perf please

**Comment by richardalow** [comment]
@swift-nio-bot test perf please

**Comment by Lukasa** [comment]
Lol this made those benchmarks pretty fast. We should follow up by increasing the iteration counts of those tests, probably.

**Comment by richardalow** [comment]
@Lukasa PR is feature complete, please let me know if more changes are required

**Comment by weissi** [comment]
CC @tanner0101 / @pushkarnk : @richardalow gives everybody some pretty awesome WebSocket perf wins here!

File: Sources/NIOPerformanceTester/ByteBufferBenchmark.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+
+    func tearDown() {
+    }
+
+    func run() -> Int {
+        for _ in 1...self.iterations {
+            self.buffer.writeBytes(self.circularBuffer)
+            self.buffer.clear()
```
Comment (Lukasa) [inline Sources/NIOPerformanceTester/ByteBufferBenchmark.swift]: We can remove this `clear` call.

**Comment by richardalow** [comment]
> CC @tanner0101 / @pushkarnk : @richardalow gives everybody some pretty awesome WebSocket perf wins here!

Don't get too excited about this - I think the improvement in the benchmark is because of a speedup in the ByteBuffer allocator used in the test. The writeBytes with Repeated<UInt8> was very slow so allowing specialization has sped it up, but I can't detect any other saving.

Anyway, this shows how this PR can help other types significantly.

**Comment by Lukasa** [comment]
Hah, yes, good spot!

**Pull request merged**